### PR TITLE
repo: add commit hooks

### PR DESCRIPTION
**Pull request by perf2711**
This PR adds two commit hooks:
* `pre-commit` - formats all modified files using `prettier` and lints them using `eslint`
* `commit-msg` - checks if the commit message adheres to `prefix: message` format

### CODE REVIEW

File: .husky/scripts/commit-msg.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    console.error(chalk.red(repeat(str.length + 10, '=')));
+}
+
+const message = process.argv[2];
+const modifiedFiles = execSync('git diff --cached --name-only').toString('utf-8');
+const expected = getExpectedPrefixes(modifiedFiles);
+const [actual, rest] = getActualPrefixes(message);
+
+if (!actual.length) {
+    banner('Commit message is invalid - missing prefix');
```
Comment (konraddysput) [inline .husky/scripts/commit-msg.js:134]: still makes sense to add an NFC commit - for example: how you can modify README.md in the root directory?

**Review by konraddysput — CHANGES_REQUESTED** [review]
I think we also need to introduce some guide for this - for example: how to extend the list. We also need to create a CONTRIBUTE guide that explains what we require from commit messages and when

File: .husky/scripts/commit-msg.js (hunk 1 above)
Comment (perf2711) [reply .husky/scripts/commit-msg.js:134]: This only will detect changes in `packages/`, `tools/`, `smoketests/` and `scripts/` directories.

File: .husky/scripts/commit-msg.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+ */
+const ciEquals = (a, b) => a.localeCompare(b, undefined, { sensitivity: 'accent' }) === 0;
+
+/**
+ * Expected change prefixes, in form of `[predicate, getPrefix]` functions.
+ */
+const changePrefix = [
+    // packages/sdk-core/file.ts => sdk-core
+    // tools/cli/file.ts => cli
+    [oneOf(startsWith('packages/'), startsWith('tools/')), nthPathPart(1)],
```
[the remark this answers is not in the document]
Comment (perf2711) [reply .husky/scripts/commit-msg.js:39]: I don't understand the question.
The comment above explains what is expected:
```
// packages/sdk-core/file.ts => sdk-core
// tools/cli/file.ts => cli
```

File: .husky/pre-commit [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env sh
+
+# Format modified files
+prettier $(git diff --cached --name-only --diff-filter=ACMR | grep -e ".ts$" | sed 's| |\\ |g') --write --ignore-unknown
+
+# Update index with modified files
+git update-index --again
```
[the remark this answers is not in the document]
Comment (perf2711) [reply .husky/pre-commit:7]: I figured it should pick up the changes made by Prettier.

**Comment by perf2711** [comment]
>I think we also need to introduce some guide for this - for example: how to extend the list.

I added some docs to `commit-msg.js` file.

>We also need to create a CONTRIBUTE guide that explains what we require from commit messages and when

That's out of scope of this pull request.

**Review by konraddysput — APPROVED** [review]

**Pull request merged**